serve/filters/filter/tx.go

```diff
@@ -0,0 +1,203 @@
+	changes := make([]*types.TxResult, len(tf.txrs))
+	copy(changes, tf.txrs)
+
+	tf.txrs = tf.txrs[:0] // reset for new transactions
+
+	return changes
+}
+
+// UpdateWithTx adds a transaction to the filter.
+func (tf *TxFilter) UpdateWithTx(txr *types.TxResult) {
+	tf.Lock()
+	defer tf.Unlock()
+
+	tf.txrs = append(tf.txrs, txr)
+}
+
+// ClearConditions resets the previously set conditions from the filter.
+func (tf *TxFilter) ClearConditions() *TxFilter {
+	tf.Lock()
+	defer tf.Unlock()
+
+	tf.conditions = nil
+
+	return tf
+}
+
+// Height sets a filter for the height of the transactions.
+//
+// It appends a height-based condition to the conditions slice.
+func (tf *TxFilter) Height(height int64) *TxFilter {
+	cond := condition{
+		func(txr *types.TxResult) bool {
+			return txr.Height == height
+		},
+		HeightPriority,
+	}
+	tf.insertConditionInOrder(cond)
+
+	return tf
+}
+
+// Index sets a filter for the index of the transactions.
+func (tf *TxFilter) Index(index uint32) *TxFilter {
+	cond := condition{
+		func(txr *types.TxResult) bool {
+			return txr.Index == index
+		},
+		IndexPriority,
+	}
+	tf.insertConditionInOrder(cond)
+
+	return tf
+}
+
+// GasUsed sets a filter for the gas used by transactions.
+func (tf *TxFilter) GasUsed(min, max int64) *TxFilter {
+	cond := condition{
+		func(txr *types.TxResult) bool {
+			return txr.Response.GasUsed >= min && txr.Response.GasUsed <= max
+		},
+		GasUsedPriority,
+	}
+	tf.insertConditionInOrder(cond)
+
+	return tf
+}
+
+// GasWanted sets a filter for the gas wanted by transactions.
+func (tf *TxFilter) GasWanted(min, max int64) *TxFilter {
+	cond := condition{
+		func(txr *types.TxResult) bool {
+			return txr.Response.GasWanted >= min && txr.Response.GasWanted <= max
+		},
+		GasWantedPriority,
+	}
+	tf.insertConditionInOrder(cond)
+
+	return tf
+}
+
+// Apply applies all added conditions to the transactions in the filter.
+//
+// It returns a slice of `TxResult` that satisfy all the conditions.
+func (tf *TxFilter) Apply() []*types.TxResult {
+	tf.Lock()
+	defer tf.Unlock()
+
+	if len(tf.conditions) == 0 {
+		return tf.txrs
+	}
+
+	var filtered []*types.TxResult
+
+	for _, txr := range tf.txrs {
+		pass := true
+
+		for _, condition := range tf.conditions {
+			if !condition.filter(txr) {
+				pass = false
+
+				break
+			}
+		}
+
+		if pass {
+			filtered = append(filtered, txr)
+		}
+	}
+
+	return filtered
+}
+
+// insertConditionInOrder adds a new condition to the conditions slice.
+//
+// It places the condition at the right position based on its priority,
+// ensuring the slice is always ordered without needing to sort it entirely each time.
+//
+// complexity: O(log n) + O(n)
+func (tf *TxFilter) insertConditionInOrder(cond condition) {
+	tf.Lock()
+	defer tf.Unlock()
+
+	i := sort.Search(len(tf.conditions), func(i int) bool {
+		return tf.conditions[i].priority >= cond.priority
+	})
+
+	tf.conditions = append(tf.conditions[:i], append([]condition{cond}, tf.conditions[i:]...)...)
+}
```

Comment:
You can use the package we already import - `github.com/madz-lab/insertion-queue`, for the insertion queue implementation

It's faster than using / managing these slices manually (it's a priority queue that utilizes insertion sort)